### ISSUE TITLE
CacheFilter integration test timing issue?

### ISSUE THREAD

**Issue by asraa** [envoyproxy/envoy#16587]
*Cache Filter*: *Question about tests*

*Description*:
https://github.com/envoyproxy/envoy/pull/15814

Hi, I would appreciate some help on debugging the cache filter integration test to unblock my new parser PR above. There shouldn't be any functional changes, and it doesn't appear that's the case to me immediately. 

The `ServeHeadFromCacheAfterGetRequest` https://github.com/envoyproxy/envoy/blob/94d11374c57dcfb32cdb6f61bb93b64784cb4b34/test/extensions/filters/http/cache/cache_filter_integration_test.cc#L318 test is failing with the new parser in the PR reliably:

```
bazel test //test/extensions/filters/http/cache:cache_filter_integration_test  --define use_new_http1_parser_in_integration_tests=true --test_output=all --test_filter=Protocols/CacheIntegrationTest.ServeGetFromUpstreamAfterHeadRequest/IPv4_HttpDownstream_HttpUpstream --test_arg="-l trace"
```

It seems like for some reason, the response to the GET request isn't completing . Here's the different stack traces that look relevant, from the response onwards 
Good:
```
[2021-05-19 18:44:36.087][29][debug][http] [source/common/http/conn_manager_impl.cc:1472] [C1][S1305557876380902528] encoding headers via codec (end_stream=false):
':status', '200'
'date', 'Thu, 01 Jan 1970 01:00:00 GMT'
'cache-control', 'public,max-age=3600'
'content-length', '42'
'x-envoy-upstream-service-time', '0'
'server', 'envoy'

[2021-05-19 18:44:36.087][29][trace][connection] [source/common/network/connection_impl.cc:470] [C1] writing 161 bytes, end_stream false
[2021-05-19 18:44:36.087][29][trace][http] [source/common/http/http1/codec_impl.cc:822] [C2] message complete
[2021-05-19 18:44:36.087][29][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x5d257f731100 for 300000ms, min is 300000ms
[2021-05-19 18:44:36.087][29][debug][cache_filter] [source/extensions/filters/http/cache/cache_filter.cc:127] [C1][S1305557876380902528] CacheFilter::encodeData inserting body
[2021-05-19 18:44:36.088][29][trace][http] [source/common/http/filter_manager.cc:1205] [C1][S1305557876380902528] encode data called: filter=0x5d257f750000 status=0
[2021-05-19 18:44:36.088][29][trace][http] [source/common/http/conn_manager_impl.cc:1481] [C1][S1305557876380902528] encoding data via codec (size=42 end_stream=false)
[2021-05-19 18:44:36.088][29][trace][connection] [source/common/network/connection_impl.cc:470] [C1] writing 42 bytes, end_stream false
[2021-05-19 18:44:36.088][29][trace][http] [source/common/http/http1/codec_impl.cc:1347] [C2] message complete
[2021-05-19 18:44:36.088][29][debug][client] [source/common/http/codec_client.cc:130] [C2] response complete
[2021-05-19 18:44:36.088][29][trace][main] [source/common/event/dispatcher_impl.cc:252] item added to deferred deletion list (size=1)
[2021-05-19 18:44:36.089][29][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x5d257f731100 for 300000ms, min is 300000ms
[2021-05-19 18:44:36.089][29][debug][cache_filter] [source/extensions/filters/http/cache/cache_filter.cc:127] [C1][S1305557876380902528] CacheFilter::encodeData inserting body
[2021-05-19 18:44:36.089][29][trace][http] [source/common/http/filter_manager.cc:1205] [C1][S1305557876380902528] encode data called: filter=0x5d257f750000 status=0
[2021-05-19 18:44:36.089][29][trace][http] [source/common/http/conn_manager_impl.cc:1481] [C1][S1305557876380902528] encoding data via codec (size=0 end_stream=true)
[2021-05-19 18:44:36.089][29][trace][connection] [source/common/network/connection_impl.cc:349] [C1] readDisable: disable=false disable_count=1 state=0 buffer_length=0
[2021-05-19 18:44:36.089][29][trace][main] [source/common/event/dispatcher_impl.cc:252] item added to deferred deletion list (size=2)
[2021-05-19 18:44:36.089][29][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x5d257f731180 for 3600000ms, min is 3600000ms
[2021-05-19 18:44:36.089][29][debug][pool] [source/common/http/http1/conn_pool.cc:53] [C2] response complete
[2021-05-19 18:44:36.089][29][debug][pool] [source/common/conn_pool/conn_pool_base.cc:200] [C2] destroying stream: 0 remaining
```

Bad (test eventually times out, response does not end stream and complete)
```
[2021-05-19 18:49:21.841][29][debug][http] [source/common/http/conn_manager_impl.cc:1472] [C1][S6494136048566019688] encoding headers via codec (end_stream=false):
':status', '200'
'date', 'Thu, 01 Jan 1970 01:00:00 GMT'
'cache-control', 'public,max-age=3600'
'content-length', '42'
'x-envoy-upstream-service-time', '0'
'server', 'envoy'

[2021-05-19 18:49:21.842][29][trace][connection] [source/common/network/connection_impl.cc:470] [C1] writing 161 bytes, end_stream false
[2021-05-19 18:49:21.842][29][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x7d53f69f980 for 300000ms, min is 300000ms
[2021-05-19 18:49:21.842][29][debug][cache_filter] [source/extensions/filters/http/cache/cache_filter.cc:127] [C1][S6494136048566019688] CacheFilter::encodeData inserting body
[2021-05-19 18:49:21.842][29][trace][http] [source/common/http/filter_manager.cc:1205] [C1][S6494136048566019688] encode data called: filter=0x7d53f710000 status=0
[2021-05-19 18:49:21.842][29][trace][http] [source/common/http/conn_manager_impl.cc:1481] [C1][S6494136048566019688] encoding data via codec (size=42 end_stream=false)
[2021-05-19 18:49:21.842][29][trace][connection] [source/common/network/connection_impl.cc:470] [C1] writing 42 bytes, end_stream false
[2021-05-19 18:49:21.842][29][trace][http] [source/common/http/http1/codec_impl.cc:612] [C2] parsed 154 bytes
[2021-05-19 18:49:21.842][29][trace][connection] [source/common/network/connection_impl.cc:547] [C1] socket event: 2
[2021-05-19 18:49:21.842][29][trace][connection] [source/common/network/connection_impl.cc:656] [C1] write ready
[2021-05-19 18:49:21.842][15][trace][connection] [source/common/network/connection_impl.cc:547] [C0] socket event: 3
[2021-05-19 18:49:21.842][15][trace][connection] [source/common/network/connection_impl.cc:656] [C0] write ready
[2021-05-19 18:49:21.842][15][trace][connection] [source/common/network/connection_impl.cc:585] [C0] read ready. dispatch_buffered_data=false
[2021-05-19 18:49:21.842][15][trace][connection] [source/common/network/raw_buffer_socket.cc:24] [C0] read returns: 203
[2021-05-19 18:49:21.842][15][trace][connection] [source/common/network/raw_buffer_socket.cc:38] [C0] read error: Resource temporarily unavailable
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:563] [C0] parsing 203 bytes
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:844] [C0] message begin
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:483] [C0] completed header: key=date value=Thu, 01 Jan 1970 01:00:00 GMT
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:483] [C0] completed header: key=cache-control value=public,max-age=3600
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:483] [C0] completed header: key=content-length value=42
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:483] [C0] completed header: key=x-envoy-upstream-service-time value=0
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:695] [C0] onHeadersCompleteBase
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:483] [C0] completed header: key=server value=envoy
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:1265] [C0] status_code 200
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:1275] [C0] Client: onHeadersComplete size=5
[2021-05-19 18:49:21.842][15][trace][http] [source/common/http/http1/codec_impl.cc:612] [C0] parsed 203 bytes
[2021-05-19 18:49:21.842][29][trace][connection] [source/common/network/raw_buffer_socket.cc:67] [C1] write returns: 203
-- Test timed out at 2021-05-19 18:54:18 UTC --
```

I'm trying to trace down what's happening in the bad stack trace. The 42 sized response body is sent with end stream false but then never completed with an end stream true.

Any leads I could go after? @ekiziv

**Comment by antoniovicente** [comment]
I'm still debugging but at first glance it seems like an error handling HEAD requests.  It seems like llhttp is not handling the NoBody return value from the header callback correctly.

**Comment by antoniovicente** [comment]
onMessageComplete is never called on the upstream connection by the parser.  It's not clear why.  I think that this happens even if you disable the cache filter.  Is it possible that the parser is not being fully reset in some cases and the request sequence of a HEAD request followed by GET on the same upstream connection causes problems?

**Comment by asraa** [comment]
Hmmm interesting, thank you so much! Let me try a repro without cache filter in HttpIntegrationTest

**Comment by antoniovicente** [comment]
@ekiziv @yanavlasov @jmarantz @htuch How can we make progress in tracking down this test failure?

**Comment by ekiziv** [comment]
I have chatted with asraa off line a couple of days ago after I spent some time looking without considerable success. @asraa, have you managed to reproduce the issue without cache filter? I am happy to keep digging, especially if you have an idea of where I should focus my attention!

**Comment by asraa** [comment]
Yeah! It seems to reproduce without adding the filter in the chain, so I'll close this as is for now.

Thank you so much @ekiziv!
